In [1]:
%load_ext nb_black
from IPython.display import display, HTML

display(HTML("<style>.container { width:100% !important; }</style>"))

<IPython.core.display.Javascript object>

In [2]:
CLASSIFICATIONS = {
    "01": "Agriculture,Forestry and Fishing",
    "02": "Agriculture, Forestry and Fishing",
    "03": "Agriculture, Forestry and Fishing",
    "05": "Mining and Quarrying",
    "06": "Mining and Quarrying",
    "07": "Mining and Quarrying",
    "08": "Mining and Quarrying",
    "09": "Mining and Quarrying",
    "10": "Manufacturing",
    "11": "Manufacturing",
    "12": "Manufacturing",
    "13": "Manufacturing",
    "14": "Manufacturing",
    "15": "Manufacturing",
    "16": "Manufacturing",
    "17": "Manufacturing",
    "18": "Manufacturing",
    "19": "Manufacturing",
    "20": "Manufacturing",
    "21": "Manufacturing",
    "22": "Manufacturing",
    "23": "Manufacturing",
    "24": "Manufacturing",
    "25": "Manufacturing",
    "26": "Manufacturing",
    "27": "Manufacturing",
    "28": "Manufacturing",
    "29": "Manufacturing",
    "30": "Manufacturing",
    "31": "Manufacturing",
    "32": "Manufacturing",
    "33": "Manufacturing",
    "35": "Electricity, Gand, Steam and Air Conditioning Supply",
    "36": "Water Supply; Sewerage, Waste Management and Remediation Activities",
    "37": "Water Supply; Sewerage, Waste Management and Remediation Activities",
    "38": "Water Supply; Sewerage, Waste Management and Remediation Activities",
    "39": "Water Supply; Sewerage, Waste Management and Remediation Activities",
    "41": "Construction",
    "42": "Construction",
    "43": "Construction",
    "45": "Wholesale and Retail Trade; Repair of Motor Vehicles and Motorcycles",
    "46": "Wholesale and Retail Trade; Repair of Motor Vehicles and Motorcycles",
    "47": "Wholesale and Retail Trade; Repair of Motor Vehicles and Motorcycles",
    "49": "Transportation and Storage",
    "50": "Transportation and Storage",
    "51": "Transportation and Storage",
    "52": "Transportation and Storage",
    "53": "Transportation and Storage",
    "55": "Accommodation and Food Service Activities",
    "56": "Accommodation and Food Service Activities",
    "58": "Information and Communication",
    "59": "Information and Communication",
    "60": "Information and Communication",
    "61": "Information and Communication",
    "62": "Information and Communication",
    "63": "Information and Communication",
    "64": "Financial and Insurance Activities",
    "65": "Financial and Insurance Activities",
    "66": "Financial and Insurance Activities",
    "68": "Real Estate Activities",
    "69": "Professional, Scientific and Technical Activities",
    "70": "Professional, Scientific and Technical Activities",
    "71": "Professional, Scientific and Technical Activities",
    "72": "Professional, Scientific and Technical Activities",
    "73": "Professional, Scientific and Technical Activities",
    "74": "Professional, Scientific and Technical Activities",
    "75": "Professional, Scientific and Technical Activities",
    "77": "Administrative and Support Service Activities",
    "78": "Administrative and Support Service Activities",
    "79": "Administrative and Support Service Activities",
    "80": "Administrative and Support Service Activities",
    "81": "Administrative and Support Service Activities",
    "82": "Administrative and Support Service Activities",
    "84": "Public Administration and Defence; Compulsory Social Security",
    "85": "Education",
    "86": "Human Health and Social Work Activities",
    "87": "Human Health and Social Work Activities",
    "88": "Human Health and Social Work Activities",
    "90": "Arts, Entertainment and Recreation",
    "91": "Arts, Entertainment and Recreation",
    "92": "Arts, Entertainment and Recreation",
    "93": "Arts, Entertainment and Recreation",
    "94": "Other Service Activities",
    "95": "Other Service Activities",
    "96": "Other Service Activities",
    "97": "Activities of Households and Employers",
    "98": "Activities of Households and Employers",
    "99": "Activities of Extraterritorial Organisations and Bodies",
}

<IPython.core.display.Javascript object>

In [3]:
import pandas as pd

<IPython.core.display.Javascript object>

In [4]:
df = pd.read_csv("../CH/BasicCompanyDataAsOneFile-2022-06-01.csv", low_memory=True)
df1 = pd.read_csv("../CH/BasicCompanyDataAsOneFile-2022-07-01.csv", low_memory=True)

C:\Users\VRUser2\AppData\Local\Temp\ipykernel_18304\3675709654.py:1: DtypeWarning: Columns (41,42,43,44,45,46,47,48,49,50,51,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../CH/BasicCompanyDataAsOneFile-2022-06-01.csv", low_memory=True)
C:\Users\VRUser2\AppData\Local\Temp\ipykernel_18304\3675709654.py:2: DtypeWarning: Columns (39,40,41,42,43,44,45,46,47,48,49,50,51,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df1 = pd.read_csv("../CH/BasicCompanyDataAsOneFile-2022-07-01.csv", low_memory=True)


<IPython.core.display.Javascript object>

In [5]:
print(df.shape)
print(df1.shape)

(5107631, 55)
(5112748, 55)


<IPython.core.display.Javascript object>

In [15]:
5112748 - 5107631

5117

<IPython.core.display.Javascript object>

In [6]:
print(df["CompanyStatus"].value_counts() * 100 / df.shape[0])
print("\n")
print(df1["CompanyStatus"].value_counts() * 100 / df.shape[0])

Active                                              92.876208
Active - Proposal to Strike off                      5.032059
Liquidation                                          1.990453
In Administration                                    0.047772
Live but Receiver Manager on at least one charge     0.025315
Voluntary Arrangement                                0.013059
In Administration/Administrative Receiver            0.005658
RECEIVERSHIP                                         0.003720
ADMINISTRATION ORDER                                 0.002682
ADMINISTRATIVE RECEIVER                              0.002114
In Administration/Receiver Manager                   0.000744
RECEIVER MANAGER / ADMINISTRATIVE RECEIVER           0.000196
VOLUNTARY ARRANGEMENT / ADMINISTRATIVE RECEIVER      0.000020
Name: CompanyStatus, dtype: float64


Active                                              92.953680
Active - Proposal to Strike off                      5.051735
Liquidation                     

<IPython.core.display.Javascript object>

In [7]:
df["Sector1"] = df["SICCode.SicText_1"].str[:2].map(CLASSIFICATIONS)
df["Sector2"] = df["SICCode.SicText_2"].str[:2].map(CLASSIFICATIONS)
df["Sector3"] = df["SICCode.SicText_3"].str[:2].map(CLASSIFICATIONS)
df["Sector4"] = df["SICCode.SicText_4"].str[:2].map(CLASSIFICATIONS)

df1["Sector1"] = df1["SICCode.SicText_1"].str[:2].map(CLASSIFICATIONS)
df1["Sector2"] = df1["SICCode.SicText_2"].str[:2].map(CLASSIFICATIONS)
df1["Sector3"] = df1["SICCode.SicText_3"].str[:2].map(CLASSIFICATIONS)
df1["Sector4"] = df1["SICCode.SicText_4"].str[:2].map(CLASSIFICATIONS)

<IPython.core.display.Javascript object>

In [8]:
print(
    df[
        (df["Sector1"] == "Manufacturing")
        | (df["Sector2"] == "Manufacturing")
        | (df["Sector3"] == "Manufacturing")
        | (df["Sector4"] == "Manufacturing")
    ].shape
)

print(
    df1[
        (df1["Sector1"] == "Manufacturing")
        | (df1["Sector2"] == "Manufacturing")
        | (df1["Sector3"] == "Manufacturing")
        | (df1["Sector4"] == "Manufacturing")
    ].shape
)

(231246, 59)
(230912, 59)


<IPython.core.display.Javascript object>

In [12]:
230912 - 231246

-334

<IPython.core.display.Javascript object>

* 334 less, compared to previous month.

In [10]:
df_m = df[
    (df["Sector1"] == "Manufacturing")
    | (df["Sector2"] == "Manufacturing")
    | (df["Sector3"] == "Manufacturing")
    | (df["Sector4"] == "Manufacturing")
]
print(df_m["CompanyStatus"].value_counts() * 100 / df_m.shape[0])

print("\n")

df_m1 = df1[
    (df1["Sector1"] == "Manufacturing")
    | (df1["Sector2"] == "Manufacturing")
    | (df1["Sector3"] == "Manufacturing")
    | (df1["Sector4"] == "Manufacturing")
]
print(df_m1["CompanyStatus"].value_counts() * 100 / df_m1.shape[0])

Active                                              92.054349
Active - Proposal to Strike off                      5.003762
Liquidation                                          2.722642
In Administration                                    0.137516
Voluntary Arrangement                                0.030703
In Administration/Administrative Receiver            0.019027
RECEIVERSHIP                                         0.014271
ADMINISTRATIVE RECEIVER                              0.009514
ADMINISTRATION ORDER                                 0.005189
Live but Receiver Manager on at least one charge     0.002595
In Administration/Receiver Manager                   0.000432
Name: CompanyStatus, dtype: float64


Active                                              92.048486
Active - Proposal to Strike off                      4.993677
Liquidation                                          2.736107
In Administration                                    0.142045
Voluntary Arrangement           

<IPython.core.display.Javascript object>

In [11]:
df_ac = df_m[df_m["CompanyStatus"] == "Active"]
df_ac1 = df_m1[df_m1["CompanyStatus"] == "Active"]
print(df_ac.shape[0])
print(df_ac1.shape[0])
print("diff:", df_ac1.shape[0] - df_ac.shape[0])

212872
212551
diff: -321


<IPython.core.display.Javascript object>

* 321 less Active Companies in terms of count

In [14]:
df_ac1.shape[0] - df_ac1[df_ac1["CompanyName"].isin(df_ac["CompanyName"])].shape[0]

3490

<IPython.core.display.Javascript object>

* 3490 new companies

In [38]:
fin = pd.read_csv("financial_data1.csv", low_memory=False)
fin.shape

(25419309, 14)

<IPython.core.display.Javascript object>

In [39]:
df_ac1[" CompanyNumber"] = df_ac1[" CompanyNumber"].apply(lambda x: x.zfill(8))
fin["Company Registered Number"] = fin["Company Registered Number"].apply(
    lambda x: x.zfill(8)
)

C:\Users\VRUser2\AppData\Local\Temp\ipykernel_15888\1915455366.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ac1[" CompanyNumber"] = df_ac1[" CompanyNumber"].apply(lambda x: x.zfill(8))


<IPython.core.display.Javascript object>

In [89]:
df_ac1[df_ac1[" CompanyNumber"].isin(fin["Company Registered Number"])].shape

(149721, 59)

<IPython.core.display.Javascript object>

In [98]:
df_ac1[~df_ac1[" CompanyNumber"].isin(fin["Company Registered Number"])].shape

(63151, 59)

<IPython.core.display.Javascript object>

In [41]:
149721 / df_ac1.shape[0]

0.7033381562629185

<IPython.core.display.Javascript object>

In [43]:
fin[fin["Company Registered Number"].isin(df_ac1[" CompanyNumber"])].to_csv(
    "manf_finan1.csv", header=True, index=False
)

<IPython.core.display.Javascript object>

In [105]:
from sqlalchemy import create_engine, MetaData, Table

cnx = create_engine("mysql+pymysql://root:DataisBeaut1ful!@127.0.0.1/vc_ds")
cnx.execute("""SELECT COUNT(*) FROM bulk_info;""").fetchall()

bulk_info = pd.read_sql(
    """SELECT CompanyName, CompanyNumber FROM bulk_info;""", con=cnx
)
bulk_info.shape

(485449, 2)

<IPython.core.display.Javascript object>

In [93]:
df_ac1[df_ac1["CompanyName"].isin(bulk_info["CompanyName"])].shape

(19282, 59)

<IPython.core.display.Javascript object>

In [94]:
df_ac1[df_ac1[" CompanyNumber"].isin(bulk_info["CompanyNumber"])].shape

(16070, 59)

<IPython.core.display.Javascript object>

In [83]:
df_ac1[
    (df_ac1["CompanyName"].isin(bulk_info["CompanyName"]))
    | (df_ac1[" CompanyNumber"].isin(bulk_info["CompanyNumber"]))
].shape

(19311, 59)

<IPython.core.display.Javascript object>

In [84]:
df_ac1.shape

(212872, 59)

<IPython.core.display.Javascript object>

In [46]:
manf_fin = pd.read_csv("manf_finan1-cleaned-classified.csv")
manf_fin.shape

C:\Users\VRUser2\AppData\Local\Temp\ipykernel_15888\3236086430.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  manf_fin = pd.read_csv("manf_finan1-cleaned-classified.csv")


(147345, 7)

<IPython.core.display.Javascript object>

In [47]:
manf_fin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147345 entries, 0 to 147344
Data columns (total 7 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Company Registered Number  147345 non-null  object 
 1   Equity                     147345 non-null  float64
 2   Percentage Change          147345 non-null  object 
 3   SME Classification         147345 non-null  object 
 4   Equity Category            147345 non-null  object 
 5   PCT Category               147345 non-null  object 
 6   Category                   147345 non-null  object 
dtypes: float64(1), object(6)
memory usage: 7.9+ MB


<IPython.core.display.Javascript object>

In [25]:
manf_fin["Category"].value_counts()

Strugglers           68820
Prospects            38704
No Classification    33072
Contenders            5693
Champions              954
Star Performers        188
Name: Category, dtype: int64

<IPython.core.display.Javascript object>

In [48]:
manf_fin["Category"].value_counts()

Strugglers           68623
Prospects            38800
No Classification    33025
Contenders            5738
Champions              964
Star Performers        195
Name: Category, dtype: int64

<IPython.core.display.Javascript object>

In [57]:
(2376 + 33025) / 147345

0.24025925548881877

<IPython.core.display.Javascript object>

In [49]:
manf_fin["Company Registered Number"] = manf_fin["Company Registered Number"].apply(
    lambda x: str(x).zfill(8)
)

<IPython.core.display.Javascript object>

In [50]:
df_ac1[df_ac1[" CompanyNumber"].isin(manf_fin["Company Registered Number"])].shape

(147345, 59)

<IPython.core.display.Javascript object>

In [51]:
pd.merge(
    df_ac1,
    manf_fin,
    how="left",
    left_on=" CompanyNumber",
    right_on="Company Registered Number",
).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 212872 entries, 0 to 212871
Data columns (total 66 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   CompanyName                         212872 non-null  object 
 1    CompanyNumber                      212872 non-null  object 
 2   RegAddress.CareOf                   2154 non-null    object 
 3   RegAddress.POBox                    618 non-null     object 
 4   RegAddress.AddressLine1             212870 non-null  object 
 5    RegAddress.AddressLine2            148777 non-null  object 
 6   RegAddress.PostTown                 211213 non-null  object 
 7   RegAddress.County                   87361 non-null   object 
 8   RegAddress.Country                  146209 non-null  object 
 9   RegAddress.PostCode                 212699 non-null  object 
 10  CompanyCategory                     212872 non-null  object 
 11  CompanyStatus             

<IPython.core.display.Javascript object>

In [53]:
manf_fin["Company Registered Number"].value_counts()

00019071    1
10913594    1
10913351    1
10913450    1
10913502    1
           ..
06765556    1
06765419    1
06765388    1
06765026    1
SC700594    1
Name: Company Registered Number, Length: 147345, dtype: int64

<IPython.core.display.Javascript object>

```{toggle}
* No.of **Active Manufacturing** Companies having Financial records: 149,721 (70%)
    * Strugglers - 68,623
    * Prospects - 38,800
    * No Classification - 33,025
    * Contenders - 5,738
    * Champions - 964
    * Star Performers - 195
* No.of **Active Manufacturing** Companies having no financial records: 63,151
* No.of **Active Manufacturing** Companies having Financial Category: 1,14,320 (Strugglers+Prospects+Contenders+Champions+Star Performers)
* No.of **Active Manufacturing** Companies having no Financial Category: 33,025 + 63,151 = 96,176
------------------------
* Total No.of **Active Manufacturing Companies** matched with **bulk_info**: 19,311
```

In [100]:
df = pd.read_excel("../InnovateUK/IUK-040122-FundedProjects.xlsx", engine="openpyxl")
df["CRN"] = df["CRN"].apply(lambda x: str(x).zfill(8))

<IPython.core.display.Javascript object>

In [103]:
df_ac1.drop(columns=["Sector1", "Sector2", "Sector3", "Sector4"]).to_csv(
    "June_Manf_Companies.csv", header=True, index=False
)

<IPython.core.display.Javascript object>

In [107]:
bulk_info[["CompanyName"]].to_csv("all_bulk_comps.csv", header=True, index=False)

<IPython.core.display.Javascript object>

# Summary

* No.of Companies from June: 5,107,631
* No.of Companies from Junly: 5,112,748
* diff: 5,117
-----------------------
* % .of **Active** Companies from June: 92.876208
* % .of **Active** Companies from Junly: 92.953680

--------------------
* No.of **Manufacturing** Companies from June: 231,246
* No.of **Manufacturing** Companies from Junly: 230,912
* diff: -334
------------------------
* No.of **Active Manufacturing** Companies from June: 212,872
* No.of **Active Manufacturing** Companies from Junly: 212,551
* diff: -321